langworld_db_data/adders/listed_value_adder.py

```diff
@@ -66,40 +74,48 @@ def _add_to_inventory_of_listed_values(
+        Collecting is done to get index and line number of final value (if the new value itself is intended final)
+        and to calculate line number for the new value (if it is intended non-final).
+        """
+
+        for row in rows:
+            if row["en"] == new_value_en or row["ru"] == new_value_ru:
+                raise ValueError(f"Row {row} already contains value you are trying to add")
+
         value_indices_to_inventory_line_numbers = (
-            self._get_indices_and_their_line_numbers_in_features_listed_values(
+            self._get_indices_and_their_line_numbers_for_given_feature_in_features_listed_values(
                 rows=rows,
                 feature_id=feature_id,
-                new_value_en=new_value_en,
-                new_value_ru=new_value_ru,
             )
         )
 
-        # Check if index_to_assign is -1 or belongs to range of indices in value_indices_to_inventory_line_numbers
+        # Check if passed index is valid
         last_index_in_feature = value_indices_to_inventory_line_numbers[-1]["index"]
-        acceptable_indices_to_assign = [-1] + [i for i in range(1, last_index_in_feature)]
+        acceptable_indices_to_assign = set([-1] + [i for i in range(1, last_index_in_feature + 2)])
         if index_to_assign not in acceptable_indices_to_assign:
             raise ValueError(
-                f"Invalid index_to assign (must be between 1 and {last_index_in_feature}, "
+                f"Invalid index_to assign (must be between 1 and {last_index_in_feature + 1}, "
                 f"{index_to_assign} was given)"
             )
 
         # id_of_new_value and line_number_of_new_value are assigned for default case and then changed if necessary
-        id_of_new_value = (
-            f"{feature_id}{ID_SEPARATOR}{value_indices_to_inventory_line_numbers[-1]['index'] + 1}"
-        )
-        line_number_of_new_value = value_indices_to_inventory_line_numbers[-1]["line number"] + 1
+        if index_to_assign in (-1, last_index_in_feature + 1):
+            id_of_new_value = f"{feature_id}{ID_SEPARATOR}{value_indices_to_inventory_line_numbers[-1]['index'] + 1}"
+            line_number_of_new_value = (
+                value_indices_to_inventory_line_numbers[-1]["line number"] + 1
+            )
+            rows_with_updated_value_indices = rows.copy()
 
         # If value is inserted into range of values, IDs following it must be incremented
-
-        if index_to_assign > -1:
-
+        else:
             id_of_new_value = f"{feature_id}{ID_SEPARATOR}{index_to_assign}"
 
             # Go through values of the feature, ignore indices less than index_to_assign,
             # increment all other indices
-            rows = self._increment_ids_whose_indices_are_not_less_than_index_to_assign_in_rows(
+            rows_with_updated_value_indices = self._increment_ids_whose_indices_are_not_less_than_index_to_assign_in_rows(
```

Comment:
```suggestion
            rows_with_updated_value_indices = self._increment_ids_whose_indices_are_equal_or_greater_than_index_to_assign(
```

Reply:
I know why you put `_in_rows`, but it's so far away from `increment` that a reader would barely guess what it refers to